### PR TITLE
Reduce port range re-use in tests

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we select the port range to use in tests by taking the Gradle
worker ID mod 223. We now use significantly more than 223 Gradle workers
in each test run which means port ranges are re-used, resulting in
spurious failures.

This commit replaces the magic number 223 with a computation defined in
terms of the available ports and the number of ports to allocate per
worker. It also reduces the number of ports per worker to 30 to try and
reduce the frequency of port clashes.

### CODE REVIEW

**Review by henningandersen — COMMENTED** [review]
I think this will not work well with the gradle daemon? Perhaps we can still do the modulus and assert something around the worker id in the build somewhere?

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 1]
```diff
@@ -1647,38 +1646,73 @@ public static boolean inFipsJvm() {
...
+     */
     protected static int getBasePort() {
-        // some tests use MockTransportService to do network based testing. Yet, we run tests in multiple JVMs that means
-        // concurrent tests could claim port that another JVM just released and if that test tries to simulate a disconnect it might
-        // be smart enough to re-connect depending on what is tested. To reduce the risk, since this is very hard to debug we use
-        // a different default port range per JVM unless the incoming settings override it
-        // use a non-default base port otherwise some cluster in this JVM might reuse a port
-
-        // We rely on Gradle implementation details here, the worker IDs are long values incremented by one for the
-        // lifespan of the daemon this means that they can get larger than the allowed port range.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:1665]: Darn it, yes. I've asked @elastic/es-delivery for some input here. I would hope we don't use the daemon in CI, at least we didn't seem to in the runs this PR has had so far, but it'd be good to confirm this.

[the remark this answers is not in the document]
Comment (henningandersen) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:1665]: Could we capture an offset very early in the build? I.e., capture the `initial_worker_id` and use `worker_id - initial_worker_id` for the port range?

**Comment by rjernst** [comment]
Could we just use an ephemeral port? There are only a handful of places we use `getPortRange()`, and I just replaced those with 0, and the resulting tests pass. Looking at each of these, we most of the time create a transport with the settings, then introspect the port that was bound anyways. I don't see anywhere that we take the port returned from getPortRange and try to use it, we always just pass to settings and later figure out what port was used.

**Comment by rjernst** [comment]
(Note that I understand my request is unrelated to the original issue here, it is just something I noticed while reviewing, and would at least reduce the re-use a bit).

**Comment by mark-vieira** [comment]
Chatting Slack a bit about this but wanted to share a summary here for posterity.

The concerns around the daemon are somewhat irrelevant, the core problem exists even with a single-use daemon in a single test invocation. That is, we now have a strict upper limit on the number of test workers that can get launched by a single build. Eventually we will hit this limit, and in fact, we might already as our PR checks run subsets of our entire test suite in parallel for efficiency and later on we have jobs that might fail because they run all tests in a single build execution. So essentially we're adding a timebomb to our build which we have no idea when it might go off. Not ideal.

Instead we'll go forward with this change but add back the leniency bit from before. We still get the benefit of narrower port ranges so the likelyhood of overlap reduces. We can then log when we're reusing a port range that had been used before, at least giving us some insight as to why we might be seeing funny failures.

Long term we'll need a better solution here which I've described in https://github.com/elastic/elasticsearch/issues/85786.

**Comment by DaveCTurner** [comment]
Ok, I reinstated the wraparound behaviour but left `PORTS_PER_WORKER` at 30 which should help a bunch.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**